sprocs/chunks_insert.sql

```diff
@@ -0,0 +1,28 @@
+CREATE OR REPLACE FUNCTION
+    chunks_insert(
+        IN in_course_id bigint,
+        IN generated_chunks json[]
+    ) RETURNS void
+AS $$
+BEGIN
+    INSERT INTO chunks (type, course_id, question_id, course_instance_id, assessment_id, uuid)
+    SELECT
+        (chunk->>'type')::enum_chunk_type,
+        in_course_id,
+        q.id,
+        ci.id,
+        a.id,
+        (chunk->>'uuid')::uuid
+    FROM unnest(generated_chunks) AS chunk(json)
+    LEFT JOIN questions AS q ON (q.uuid = (chunk->>'questionUuid')::uuid AND q.qid = (chunk->>'questionName') AND q.course_id = in_course_id)
+    LEFT JOIN course_instances AS ci ON (ci.uuid = (chunk->>'courseInstanceUuid')::uuid AND ci.short_name = (chunk->>'courseInstanceName') AND ci.course_id = in_course_id)
+    LEFT JOIN assessments AS a ON (a.uuid = (chunk->>'assessmentUuid')::uuid AND a.tid = (chunk->>'assessmentName') AND a.course_instance_id = ci.id)
```

Comment:
Are there any concerns about query performance when querying `uuid`? The only table with an index that *might* help here is `questions`, but even that has a `WHERE deleted_at_ IS NULL`, so I'm not sure the query engine will decide to use it.

Aside: do we need/want `deleted_at` checks here? It's late and I can't think of the answer myself, but wanted to raise it for discussion.

Reply:
This is excellent! I was looking at the DB table scan statistics the other day and I noticed that by far the worst table is `questions`, but I didn't have time to explore why. I think it's because we have an index on `(course_id, uuid) WHERE deleted_at IS NULL` but we should _not_ have the `deleted_at IS NULL` condition there. In fact UUIDs are globally unique (within a course), even among deleted rows. This is what we had in the `paper_sync.pdf` document describing the sync algorithm.

We do have the correct indexes on `course_instances` and `assessments`:
```
course_instances_course_id_uuid_key: UNIQUE (course_id, uuid) USING btree (course_id, uuid)
assessments_course_instance_id_uuid_key: UNIQUE (course_instance_id, uuid) USING btree (course_instance_id, uuid)
```

Considering `course_instances`, the LEFT JOIN condition is `ci.uuid = THE_UUID AND ci.course_id = in_course_id` so it will use the index above.

Considering `questions`, the LEFT JOIN condition is basically the same, so we would be ok if we had an index without `deleted_at IS NULL`.

I'm less certain about `assessments`. This would be fine if we were using regular JOIN rather than LEFT JOIN for `course_instances`, so we would definitely have a `ci.id`. Fortunately there are only 9000 assessments in the DB so even if it falls back to a sequential scan it will still be fast enough at the moment. I'll make an issue to check the query planner for this after we've deployed.

Reply:
I pushed a new commit to fix the index on `questions` and I added an issue to investigate the query planning after deploy.